tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/BlockingMiddleware.cs

```diff
@@ -0,0 +1,116 @@
+    {
+        var httpResponse = context.Response;
+        if (!httpResponse.HasStarted)
+        {
+            httpResponse.Clear();
+            foreach (var cookie in context.Request.Cookies)
+            {
+                httpResponse.Cookies.Delete(cookie.Key);
+            }
+
+            httpResponse.Headers.Clear();
+            httpResponse.StatusCode = 403;
+            var template = settings.BlockedJsonTemplate;
+            if (context.Request.Headers["Accept"].ToString().Contains("text/html"))
+            {
+                httpResponse.ContentType = "text/html";
+                template = settings.BlockedHtmlTemplate;
+            }
+            else
+            {
+                httpResponse.ContentType = "application/json";
+            }
+
+            endedResponse = true;
+            return httpResponse.WriteAsync(template);
+        }
+
+        context.Response.Body.Dispose();
+        endedResponse = true;
+        return Task.CompletedTask;
+    }
+
+    internal async Task Invoke(HttpContext context)
+    {
+        var security = Security.Instance;
+        var endedResponse = false;
+        if (security.Settings.Enabled)
+        {
+            var securityTransport = new SecurityTransport(security, context, Tracer.Instance.ActiveScope.Span as Span);
+            if (_runSecurity)
+            {
+                if (_endPipeline && !context.Response.HasStarted)
+                {
+                    context.Response.StatusCode = 404;
+                }
+
+                using var result = securityTransport.ShouldBlock();
+                if (result.ShouldBeReported)
+                {
+                    if (result.Block)
+                    {
+                        await WriteResponse(context, security.Settings, out endedResponse).ConfigureAwait(false);
+                        securityTransport.MarkBlocked();
+                    }
+
+                    securityTransport.Report(result, endedResponse);
+                }
+            }
+
+            if (_next != null && !endedResponse)
+            {
+                // unlikely that security is disabled and there's a blockexception, but might happen as race condition
+                try
+                {
+                    await _next(context).ConfigureAwait(false);
+                }
+                catch (BlockException e)
+                {
+                    await WriteResponse(context, security.Settings, out endedResponse).ConfigureAwait(false);
+                    if (!e.Reported)
+                    {
+                        securityTransport.Report(e.Result, endedResponse);
+                    }
+
+                    securityTransport.Cleanup();
+                }
+            }
+        }
+        else
+        {
+            if (_next != null)
+            {
+                await _next(context).ConfigureAwait(false);
+            }
```

Comment:
Do we need to catch/handle a Block Exception here? If so, we probably need something like this:

```suggestion
            try
            {
                await _next(context).ConfigureAwait(false);
            }
            catch (BlockException e)
            {
                // what should we do, as we have no securityTransport?
            }
```

If not, you can potentially improve perf slightly by moving the contents of the `if` block to another method, and removing the `await` from the outer method, e.g. something like




```csharp
internal Task Invoke(HttpContext context)
{
    return Security.Instance.Settings.Enabled
        ? CheckSecurity(context)
        : _next?.Invoke(context);
}

private async Task CheckSecurity(HttpContext context)
{
    var security = Security.Instance;
    var endedResponse = false;
    var securityTransport = new SecurityTransport(security, context, Tracer.Instance.ActiveScope.Span as Span);
    // ...
}  
```

Reply:
good point..  I guess potentially, security could be disabled by remote config, and just before a block exception could have been thrown along a previous middleware, so it would still need catching, but it seems like an unlikely race condition...

Reply:
fixed in [ef58dfb](https://github.com/DataDog/dd-trace-dotnet/pull/3442/commits/ef58dfbb0ad8fc4f3297dac3eca7eb91655b72ea)
In the end, I preferred catching it in any case as rcm could potentially disable appsec just after a request was blocked at the beginning of the lifecycle 🤯